testsuite/classlibrary/TestCoreUGens.sc

```diff
@@ -309,125 +316,143 @@ TestCoreUGens : UnitTest {
+		this.assert(testsFinished, "TIMEOUT: exact_convergence tests", report: false);
 	}
 
+
 	test_muladd {
-		var n, v;
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var n, v, tests, testDur, testsFinished;
 
-		var tests = Dictionary[
-		];
+		tests = Dictionary[];
 		[[\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5], [\ar,\kr], [2,0,5]].allTuples.do{|tup|
-			//tup.postln;
 			tests["%%.madd(%%, %%)".format(*tup)] =
 			"{DC.%(%).madd(DC.%(%), DC.%(%)) - (% * % + %)}".format(*(tup ++ tup[1,3..])).interpret;
 		};
 
 		server.bootSync;
+		testDur = defaultNumBlocksToTest * server.options.blockSize / server.sampleRate;
+
 		tests.keysValuesDo{|name, func|
-			func.loadToFloatArray(0.1, server, { |data|
+			func.loadToFloatArray(testDur, server, { |data|
 				this.assertArrayFloatEquals(data, 0, name.quote, report: true);
 				completed = completed + 1;
-				condvar.signalOne;
+				cond.signalOne;
 			});
-			rrand(0.06, 0.15).wait;
+			server.sync;
 		};
 
-		condvar.waitFor(1, { completed == tests.size });
+		testsFinished = cond.waitFor((1.5 * tests.size * testDur).max(minTimeOut), { completed == tests.size });
+		this.assert(testsFinished, "TIMEOUT: muladd tests", report: false);
 	}
 
 
-	test_bufugens{
-		var d, b, c;
+	test_bufugens {
+		var d, b, c, testsFinished;
 		var tests = [1, 2, 8, 16, 32, 33];
-		var condvar = CondVar();
+		var cond = CondVar();
 		var completed = 0;
+		var bufDur = 0.2;
 
 		server.bootSync;
 
 		// channel sizes for test:
 		tests.do{ |numchans|
-
+			var data, dataBuf, recBuf;
 			// Random data for test
-			d = {1.0.rand}.dup((server.sampleRate * 0.25).round * numchans);
-
+			data = {1.0.rand}.dup((server.sampleRate * bufDur).round * numchans);
 			// load data to server
-			b = Buffer.loadCollection(server, d, numchans);
+			dataBuf = Buffer.loadCollection(server, data, numchans);
 			// a buffer for recording the results
-			c = Buffer.alloc(server, d.size / numchans, numchans);
+			recBuf = Buffer.alloc(server, data.size / numchans, numchans);
 			server.sync;
 
 			// Copying data from b to c:
 			{
-				RecordBuf.ar(PlayBuf.ar(numchans, b, BufRateScale.ir(b), doneAction: 2), c, loop:0) * 0.1;
-			}.play(server);
-			server.sync;
-			1.0.wait;
-			c.loadToFloatArray(action: { |data|
-				// The data recorded to "c" should be exactly the same as the original data "d"
-				this.assertArrayFloatEquals(data - d, 0,
-					"data->loadCollection->PlayBuf->RecordBuf->loadToFloatArray->data (% channels)".format(numchans), report: true);
-				b.free;
-				c.free;
-				completed = completed + 1;
-				condvar.signalOne;
-			});
-			0.32.wait;
+				RecordBuf.ar(PlayBuf.ar(numchans, dataBuf, BufRateScale.ir(dataBuf), doneAction: 2), recBuf, loop:0) * 0.1;
+			}.play(server, outbus: server.options.numInputBusChannels + server.options.numOutputBusChannels); // avoid hw output: write to first virutal bus
+
 			server.sync;
+
+			fork {
```

Comment:
Would be good to have feedback on this: a routine is embedded within the test. I _think_ this is ok because the timing condition is waiting for the `completed` variable update, which is shared between the main routine and this forked routine.

Reply:
yes, it should be ok. A `forkIfNeeded` would embed the test if it happens in  a routine already.

Reply:
actually, I was wrong, better *not* use a forkIfNeeded.

Reply:
```

// works as expected
(
fork {

	var con = CondVar.new;
	fork {
		2.wait;
		con.signalOne;
	};
	con.wait;
	"hello".postln
	
}
)


// doesn't work correctly
(
fork {

	var con = CondVar.new;
	forkIfNeeded {
		2.wait;
		con.signalOne;
	};
	con.wait;
	"hello".postln
	
}
)
```

Reply:
Yes, you can think of the routine being "flattened" to remove the inner fork:
```supercollider
// doesn't work correctly
(
fork {

	var con = CondVar.new;
	forkIfNeeded {
		"waiting".postln;
		2.wait;
		"waited, signalling".postln;
		con.signalOne;
	};
	"waiting outside".postln;
	con.wait;
	"hello".postln
}
)
```
So it signals the condition before the thread is made to wait on the condition.

Reply:
yes, we could mention this in `forkIfNeeded` sometime.